### DISCUSSION TITLE
Alarms implementation with Milo

### DISCUSSION THREAD

**Discussion by hermotech** [eclipse-milo/milo#899] [General]
After a fair bit of research, it appears that Milo does not support alarms out of the box?
What I mean with out of the box is as follows:
If I set up a native OPC UA alarm, and I trigger it (for example changing a variable node that serves as the source of a condition node (alarm), I would expect to receive event notifications on my event subscription.

With node-opcua I achieved just this with minimal wiring:
I create a variable node called tankLevel (double)
Then I create a NonExclusiveLimitAlarm called tankLevelCondition and I set its sourceNode property to point to the tankLevel variable node. Then I set the values of the four threshold levels in the alarm (lowlow, low, high, highhigh).
If I set the value of the tankLevel variable node to a value higher than the high limit for example, an event notification is triggered.
I can see this in UA Expert in the Event View and also the Alarms tab.

I cannot get this to work in Milo with the same setup as described above.
I also looked in the ExampleNamespace.java file at the startBogusEventNotifier method to see how to generate events.
I tried to modify this to create an event for a specific alarm type (NonExclusiveLimitAlarm), instead of the general BaseEventType, but I cannot see these as alarms in UA Expert. 
I am probably missing something or doing something wrong, or perhaps Milo does not support alarms?

I hope someone could give me some tips / hints so that I can get alarms working, either automatically or with some extra custom wiring.

**Comment by kevinherron** [comment]
My reply from the mailing list for posterity:

The bare minimum needed to sent Events is implemented, as you may have seen in the ExampleNamespace, but Alarms and Conditions are not implemented and not currently planned.

It should be possible to implement Alarms and Conditions yourself with probably considerable effort since it’s built on top of Event support.

I need to further look into the references you mentioned when I’m back from holiday and have access to my computer on the OPC UA specs. I think the HasCondition reference may simply be missing from the BuiltinReferenceType enum in Milo.
